src/main/java/apoc/schema/Schemas.java

```diff
@@ -99,49 +99,100 @@ public Boolean constraintExistsOnRelationship(@Name("type") String type, @Name("
+            List<String> keys = new ArrayList<>();
+            definition.getPropertyKeys().forEach(keys::add);
+
+            AssertSchemaResult info = new AssertSchemaResult(label, keys);
+            if(indexes.containsKey(label)) {
+                if (keys.size() > 1) {
+                    indexes.get(label).remove(keys);
+                } else if (keys.size() == 1) {
+                    indexes.get(label).remove(keys.get(0));
+                } else
+                    throw new IllegalArgumentException("Label given with no keys.");
+            }
 
-            Boolean alreadyDropped = false;
-            for (String key : definition.getPropertyKeys()) {
+            if (dropExisting) {
+                definition.drop();
+                info.dropped();
+            }
+
+            result.add(info);
+        }
 
-                AssertSchemaResult info = new AssertSchemaResult(label, key);
+        if (dropExisting)
+            indexes = copyMapOfObjects(indexes0);
 
-                if (!indexes.containsKey(label) || !indexes.get(label).remove(key)) {
-                    if (dropExisting && !alreadyDropped) {
-                        definition.drop();
-                        alreadyDropped = true;
-                    }
-                    if(alreadyDropped)
-                        info.dropped();
+        for (Map.Entry<String, List<Object>> index : indexes.entrySet()) {
+            for (Object key : index.getValue()) {
+                if (key instanceof String) {
+                    result.add(createSinglePropertyIndex(schema, index.getKey(), (String) key));
+                } else if (key instanceof List) {
+                    result.add(createCompoundIndex(index.getKey(), (List<String>) key));
                 }
-                result.add(info);
             }
         }
+        return result;
+    }
 
-        for (Map.Entry<String, List<String>> index : indexes.entrySet()) {
-            for (String key : index.getValue()) {
-                schema.indexFor(label(index.getKey())).on(key).create();
-                result.add(new AssertSchemaResult(index.getKey(), key).created());
-            }
+    private AssertSchemaResult createSinglePropertyIndex(Schema schema, String lbl, String key) {
+        schema.indexFor(label(lbl)).on(key).create();
+        return new AssertSchemaResult(lbl, key).created();
+    }
+
+    private AssertSchemaResult createCompoundIndex(String label, List<String> keys) {
+        db.execute(String.format("CREATE INDEX ON :%s (%s)", label, String.join(",", keys)));
+        return new AssertSchemaResult(label, keys).created();
+    }
+
+    private Boolean compareKeys(List<String> keys1, List<Object> keys2) {
+        if (keys1 == null && keys2 == null)
+            return true;
+
+        if (keys1 == null || keys2 == null)
+            return false;
+
+        if (keys1.size() != keys2.size())
+            return false;
+
+        for (String item : keys1) {
```

Comment:
what about duplicates in list 2?
better to compare both sets?

Reply:
Duplicates would be caught by the `size()` call - if the 2 arrays are of equal length with list2 having duplicates then the `for` loop would fail as an item in `keys1` wouldn't be in `keys2` - or do you mean you want to allow duplicates?

Reply:
list a:
a,b,c,a
list b
a,c,c,b

But probably doesn't matter.

Reply:
Ahh I see duplicates in both, ok, and how would you compare both sets?

Reply:
Don't worry - got it!